// mwgame/Thinkers/MissileThinker.js
var O2 = require('o2');
var Thinker = require('./MoverThinker.js');
var u = require('mwgame/Utils.js');
var CONST = require('mwgame/data/consts.js');


O2.extendClass('MissileThinker', MoverThinker, {
	
	_oOwner: null,
	_bInCannon: true,

	__construct: function() {
		O2.parent(this);
	},
	
	thinkMove: function() {
		O2.parent(this);
	},
		
	setOwner: function(oEntity) {
		this._oOwner = oEntity;
	},

	getOwner: function() {
		return this._oOwner;
	},
	
	bump: function(bx, by) {
		var m = this.oEntity;
		m.rollbackXY();
		this.oGame.missileHitsTarget(this.oEntity, null);
	},
	
	/**
	 * Test si le mobile spécifié entre en collision avec un autre mobile
	 * @return mobile percuté, ou null si aucune collision
	 */
	computeCollision : function() {
		var oEntity = this;
		var bInCannon = this._bInCannon;
		var oBumper = O2.parent(this, bInCannon ? 0 : 4);
		if (oBumper == this.getOwner() && this._bInCannon) {
			return oEntity._oEntityCollision = null;
		} else {
			this._bInCannon = false;
		}
		return oEntity._oEntityCollision = oBumper;
	},

	resolveCollision : function(oTarget) {
		this.oGame.missileHitsTarget(this.oEntity, oTarget);
	}
});

module.exports = MissileThinker;
