// mwgame/Thinkers/ShooterThinker.js
/**
 * Un thinker de base pour toute les créatures avec une IA
 */
var O2 = require('o2');
var CreatureThinker = require('./CreatureThinker.js');
var u = require('mwgame/Utils.js');

var ShooterThinker = O2.extendClass(CreatureThinker, {
	
	nShots: 0, // number of shots for the next attack
	

	// Example : init -> branching on another thinker
	thinkInit: function() {
		this.setThink('RandomWalk');
	},

	// 1 - walk in a random direction for 30 units of time
	// 2 - seek any foe
	//   - if enemy then goto attack mode
	//   - else goto 1
	// 3 - attack mode:
	// 4 - aim the target and fire 3 times
	// 5 - goto 1

	// Example : MODE RANDOM WALK & WANDER
	thinkRandomWalk_enter: function() {
		this.setActionTime(30);
		var oEntity = this.oEntity;
		oEntity.a = u.rand(0, 7) * Math.PI / 4;
		oEntity.setSpeedVector(oEntity.a, oEntity.oBlueprint.speed);
	},

	thinkRandomWalk: function() {
		this.thinkMove();
		if (this.testActionTimeOut()) {
			this.setThink('Wait30');
		}
	},
	
	thinkRandomWalk_exit: function() {
		this.actionStop();
	},
	
	// WAIT MODE 30
	thinkWait30_enter: function() {
		this.setActionTime(30);
	},
	
	thinkWait30: function() {
		if (this.testHostileVisible()) {
			this.setThink('Attack');
		}
		if (this.testActionTimeOut()) {
			this.setThink('RandomWalk');
		}
	},
	
	
	
	// ATTACK MODE 
	thinkAttack_enter: function() {
		this.actionAim(); // we aim
		this.nShots = u.rand(2, 4); // ready for 2 - 4 shots
		this.setActionTime(8); // aiming for 8 units of time
	},
	
	thinkAttack: function() {
		if (this.testActionTimeOut()) {
			this.setThink('Shoot'); // just before shooting
		}
	},
	
	thinkShoot_enter: function() {
		this.actionAim(); // we aim a last time
		this.actionAttack(); // we shoot
		this.setActionTime(15); // 15 unit of time of recovery
	},
	
	thinkShoot: function() {
		if (this.testActionTimeOut()) {
			--this.nShots; // decrease the number of shots left
			if (this.nShots > 0) { // there are shots left
				this.setThink('Shoot'); // shoot again
			} else { // no more shot left
				this.setThink('RandomWalk'); // let's walk a bit
			}
		}
	},
	
	
});

module.exports = ShooterThinker;
