// mwgame/Thinkers/Thinker.js
/** Classe de déplacement automatisé et stupidité artificielle des mobiles
 * O876 Raycaster project
 * @date 2012-01-01
 * @author Raphaël Marandet 
 * Classe de base pouvant être étendue
 * Version pour node js 2014-05-26
 */
var O2 = require('o2');
var Thinker = O2.createClass({
	oEntity: null,
	oGame: null,
	_sCurrentThinkProc: '',
	
	__construct: function() {
	},
	
	destroy: function() {
		this.oEntity = null;
		this.oGame = null;
	},
	
	setEntity: function(e) {
		this.oEntity = e;
	},
	
	getThinkProcName: function(s) {
		var sThink = s.charAt(0).toUpperCase() + s.substr(1);
		return 'think' + sThink;
	},
	
	setThink: function(s) {
		var sPrevProcName = this.getThinkProcName(this._sCurrentThinkProc);
		var sProcName = this.getThinkProcName(s);
		var sProcExit = sPrevProcName + '_exit';
		var sProcEnter = sProcName + '_enter';
		if (sProcExit in this) {
			this[sProcExit]();
		}
		if (sProcName in this) {
			this.think = this[sProcName];
		} else {
			throw new Error('Thinker action not found ' + sProcName);
		}
		this._sCurrentThinkProc = s;
		if (sProcEnter in this) {
			this[sProcEnter]();
		}
	},
	
	think: function() {}
});
module.exports = Thinker;
