// mwgame/Thinkers/WandererThinker.js
/**
 * Un thinker de base pour toute les créatures avec une IA
 */
var O2 = require('o2');
var CreatureThinker = require('./CreatureThinker.js');
var u = require('mwgame/Utils.js');

var WandererThinker = O2.extendClass(CreatureThinker, {

	// Example : init -> branching on another thinker
	thinkInit: function() {
		this.setThink('RandomWalk');
	},


	// 1 - walk in a random direction for 30 units of time
	// 2 - wait for 30 units of time

	// Example : MODE RANDOM WALK & WANDER
	thinkRandomWalk_enter: function() {
		this.setActionTime(30);
		var oEntity = this.oEntity;
		oEntity.a = u.rand(0, 7) * Math.PI / 4;
		oEntity.setSpeedVector(oEntity.a, oEntity.oBlueprint.speed);
	},

	thinkRandomWalk: function() {
		this.thinkMove();
		if (this.testActionTimeOut()) {
			this.setThink('Wait30');
		}
	},
	
	thinkRandomWalk_exit: function() {
		var oEntity = this.oEntity;
		oEntity.setSpeedVector(oEntity.a, 0);
	},
	
	// WAIT MODE 30
	thinkWait30_enter: function() {
		this.setActionTime(30);
	},
	
	thinkWait30: function() {
		if (this.testActionTimeOut()) {
			this.setThink('RandomWalk');
		}
	},
	
	
});

module.exports = WandererThinker;
