// mwgame/Items/Abstract/WandType1.js
var O2 = require('o2');
var Wand = require('./Wand.js');

/**
 * Baguette magique type 1
 * Ce type de baguette dispose de deux attaques
 * - une attaque mono-projectile lorsque le tir est instantané
 * - une attaque de N projectiles lorsque le tir est chargé
 * les projectiles n'ont aucun effets 
 * 
 * une classe qui hérite de celle ci peut personnaliser
 * - la référence des missiles tirés (defaut "p_magbolt")
 * - le cooldown de l'arme (defaut 300 ms)
 * - l'ouverture angulaire de la gerbe de missiles (lors des tirs chargés) (defaut 0.26 rad)
 * - le nombre de missiles tirés (lors des tirs chargés) (defaut 3)
 * - pour chaque missile, la liste des effets est à disposition. (pas d'effet par defaut)
 */
O2.extendClass('WandType1', Wand, {
	missileRef: 'p_magbolt',
	cooldown: 300,
	angle: 0.45,
	missiles: 3,
	chargeTime: 1333,

	
	fire: function(ctx) {
		var oInst = ctx.instance;
		var aMissiles = [];
		if (ctx.charge > this.chargeTime) {
			var fAngleBase = -this.angle / 2;
			var fAngleDiff = this.angle / (this.missiles - 1);
			for (var iMis = 0; iMis < this.missiles; ++iMis) {
				aMissiles.push(oInst.shootMissile(ctx.caster, this.missileRef, iMis * fAngleDiff + fAngleBase));
			}
		} else {
			aMissiles.push(oInst.shootMissile(ctx.caster, this.missileRef, 0));
		}
		aMissiles.forEach(function(m) {
			m.setData('effects', []);
		});
		ctx.missiles = aMissiles;
	}
});

module.exports = WandType1;
